src/assets/Generator.Shared/ArmOperationHelpers.cs

```diff
@@ -0,0 +1,389 @@
+                }
+            }
+
+            private bool IsTerminalState(out string state)
+            {
+                Response response = GetRawResponse();
+                state = string.Empty;
+                if (_info.HeaderFrom == HeaderFrom.Location)
+                {
+                    return response.Status != 202;
+                }
+
+                if (response.Status >= 200 && response.Status <= 204)
+                {
+                    if (response.ContentStream?.Length > 0)
+                    {
+                        try
+                        {
+                            using JsonDocument document = JsonDocument.Parse(response.ContentStream);
+                            foreach (JsonProperty property in document.RootElement.EnumerateObject())
+                            {
+                                if ((_info.HeaderFrom == HeaderFrom.OperationLocation ||
+                                     _info.HeaderFrom == HeaderFrom.AzureAsyncOperation) &&
+                                    property.NameEquals("status"))
+                                {
+                                    state = property.Value.GetString().ToLowerInvariant();
+                                    return s_terminalStates.Contains(state);
+                                }
+
+                                if (_info.HeaderFrom == HeaderFrom.None && property.NameEquals("properties"))
+                                {
+                                    foreach (JsonProperty innerProperty in property.Value.EnumerateObject())
+                                    {
+                                        if (innerProperty.NameEquals("provisioningState"))
+                                        {
+                                            state = innerProperty.Value.GetString().ToLowerInvariant();
+                                            return s_terminalStates.Contains(state);
+                                        }
+                                    }
+                                }
+                            }
+                        }
+                        finally
+                        {
+                            // It is required to reset the position of the content after reading as this response may be used for deserialization.
+                            response.ContentStream.Position = 0;
+                        }
+                    }
+
+                    // If provisioningState was not found, it defaults to Succeeded.
+                    if (_info.HeaderFrom == HeaderFrom.None)
+                    {
+                        return true;
+                    }
+                }
+
+                throw _clientDiagnostics.CreateRequestFailedException(response);
+            }
+        }
+
+        private static readonly string[] s_failureStates = { "failed", "canceled" };
+        private static readonly string[] s_terminalStates = s_failureStates.Append("succeeded").ToArray();
```

Comment:
Just define an array, don't need to allocate multiple things.

Reply:
Done.